Implemented the profile details page front end

MilesDev91 opened the pull request:
### What this PR does:
- Adds the Profile details page at /profiles/exampleid (will be the profile id after integration)
- Adds the request form. It is complete with validation and is set up to send the proper data to the backend.

### Screenshots / Videos:
![image](https://user-images.githubusercontent.com/34494229/150589829-a2742438-9b07-451b-9710-e564c9207da0.png)
![image](https://user-images.githubusercontent.com/34494229/150590021-5b6a8f03-f812-4818-8c09-d54b43e7091a.png)
![image](https://user-images.githubusercontent.com/34494229/150590065-8957324b-662f-4cbb-9f5e-b854a4309ede.png)


### Any information needed to test this feature:
- Navigate to /profile/exampleid to see the page. Resize screen to see responsiveness. You may have to refresh page if resizing with devtools.
- Select/Enter values in the request form to check validation. Start dates should be less than end dates, and all fields are required. Check the console logs on successful submit to check that the proper dates are passed through.

### Any issues with the current functionality:
- The page has a horizontal scrollbar on mobile. The navbar being unresponsive makes it hard to tell if it is my code that is extending the width, but it may need another pass after the navbar is fixed.
- The form has red error messages and the label turns red on invalid input, but I could not get the input border to behave the same. This is probably due to the customization I did on the date picker messing with the normal form behavior.
- The page currently does not have the extra photos below the profile details. This should be easy to implement once it is decided how we are going to implement those photos.

### Review thoughts:
- I hope that using the window.innerWidth is an acceptable way to implement responsiveness in React, it really came in handy.

closes #20

MilesDev91 committed:
Created Profile Details page

Created the ProfileDetails component and added it into a protected route in App.tsx. Created some mock data and began styling.

Fixed a syntax error in the Profile model.

MilesDev91 committed:
Request form

Began styling the right side of the page.

Added the RequestForm and FormikDatePicker components. FormikDatePicker will deconstruct the styling on the DatePicker and match the mock.

MilesDev91 committed:
Completed form

Finished the request form, complete with validation and a working submit. Styled to closely match the mocks.

Finished the custom datepicker in the request form.

Moved the handleSubmit to the RequestForm component.

MilesDev91 committed:
Styling and responsiveness

Adjusted the date picker to open no matter where you click on the input.

Styled the top part of the profile details section. Used default box-shadows from MUI

Profile details page is now responsive.

MilesDev91 committed:
Update App.tsx

Changed route to better match later implementation (and not interfere with other routes).

sundayezeilo commented on client/src/pages/ProfileDetails/RequestForm/RequestForm.tsx:
```diff
@@ -0,0 +1,85 @@
+    { startDate, endDate }: { startDate: Date; endDate: Date },
+    { setSubmitting }: FormikHelpers<{ startDate: Date; endDate: Date }>,
+  ) => {
+    console.log(startDate, endDate);
```
Let's get rid of unnecessary console logs, and ditto for all other places.

sundayezeilo commented on client/src/pages/ProfileDetails/RequestForm/RequestForm.tsx:
```diff
@@ -0,0 +1,85 @@
+  const classes = useStyles();
+
+  const initialDate = () => {
+    console.log(new Date(Date.now()), new Date(Date.now()).getHours());
```
Remove this one too

sundayezeilo reviewed (changes requested):
Good work with this one :+1:

MilesDev91 committed:
Update RequestForm.tsx

Removed some console logs

rajivtitus commented on client/src/App.tsx:
```diff
@@ -32,6 +33,7 @@ function App(): JSX.Element {
                 <ProtectedRoute exact path="/dashboard" component={Dashboard} />
                 <ProtectedRoute path="/profile/settings" component={Settings} />
                 <ProtectedRoute path="/bookings" component={Bookings} />
+                <ProtectedRoute path="/profile/exampleid" component={ProfileDetails} />
```
```suggestion
                <ProtectedRoute path="/profile/:profile-id" component={ProfileDetails} />
```
We can pass the sitters' profile id as a url param from the listings to the profile details page

MilesDev91 replied:
I believe I changed this in the integration, but I suppose there is no reason to not make this change.

rajivtitus commented on client/src/components/FormikDatePicker/FormikDatePicker.tsx:
```diff
@@ -0,0 +1,122 @@
+  [inputProps: string]: any;
+}
+
+export default function FormikDatePicker({
```
Good work with this component!

rajivtitus commented on client/src/components/FormikDatePicker/FormikDatePicker.tsx:
```diff
@@ -0,0 +1,122 @@
+interface Props {
+  dateField: string;
+  date: Date;
+  setFieldValue: (field: string, value: any) => void;
```
Won't the `value` be a Date type?

MilesDev91 replied:
Apparently datepicker passes Date | null, so I changed it to that.

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -0,0 +1,93 @@
+const boxShadow =
+  '0px 0px 1.5px rgba(0, 0, 0, 0.006),0px 0px 3.6px rgba(0, 0, 0, 0.009),0px 0px 6.8px rgba(0, 0, 0, 0.011),0px 0px 12px rgba(0, 0, 0, 0.015),0px 0px 22.1px rgba(0, 0, 0, 0.027),0px 0px 48px rgba(0, 0, 0, 0.1)';
+
+const mockProfile = {
```
👍🏽

MilesDev91 replied:
Ended up removing this in favor of MUI default box shadows

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -0,0 +1,93 @@
+
+  return (
+    <PageContainer>
+      <Grid m={window.innerWidth < 600 ? 2 : 4} justifyContent="space-around" container alignItems="flex-start">
```
Using `window.innerWidth` is fine but I'm not sure whether it works when we resize? One option would be to store these values in a state
Ref:  https://www.pluralsight.com/guides/render-window-resize-react

Also, I would recommend looking into the usage of breakpoints and using them where you can.
Ref: https://mui.com/customization/breakpoints/

MilesDev91 replied:
Ah brilliant that is the exact solution I was looking for when I was troubleshooting it. Thanks!

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -0,0 +1,93 @@
+          container
+          flexDirection="column"
+        >
+          <Typography fontSize="1.2rem" fontWeight="bold" m="3rem auto 1rem auto">
```
I would recommend setting font sizes in variants within MUI's theme and using them across the app to ensure a consistent font size

MilesDev91 replied:
Ended up breaking other pages. Should be its own ticket.

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -0,0 +1,93 @@
+          <Typography fontSize="1.2rem" fontWeight="bold" m="3rem auto 1rem auto">
+            ${mockProfile.payRate}/hr
+          </Typography>
+          <Rating sx={{ margin: 'auto' }} value={mockProfile.rating} precision={0.5} />
```
```suggestion
          <Rating sx={{ margin: 'auto' }} value={mockProfile.rating} precision={0.5} readOnly />
```

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -0,0 +1,93 @@
+            <Typography>{mockProfile.aboutMe}</Typography>
+          </Box>
+        </Grid>
+        {window.innerWidth < 600 ? <hr style={{ width: '100%' }} /> : null}
```
You can use a MUI `Divider` component in place of the HTML `hr` tag

MilesDev91 replied:
Tried this and it wasn't working, I'll try it again.

rajivtitus commented on client/src/pages/ProfileDetails/RequestForm/RequestForm.tsx:
```diff
@@ -0,0 +1,83 @@
+              dateField="startDate"
+              error={touched.startDate && Boolean(errors.startDate)}
+            />
+            <Typography sx={{ marginBottom: 1, marginTop: -1.5, color: '#d32f2f' }}>{errors.startDate}</Typography>
```
I would recommend using MUI's theme to apply colors instead

Ditto where applicable

rajivtitus commented:
@MilesDev91 Good work with this! Please have a look at the comments and address them. Additionally, I would consider refining the UI a bit (additional padding where applicable and restrict the max width of the components to preserve aspect ratio)

MilesDev91 committed:
UI updates

Used variants in Typography and standardized their font-sizes through the theme. This is subject to change.

Added error color to the theme.

Adjusted max size of the grid on different screen sizes to maintain aspect ratio. Added window width to the state and an event listener to set it so the page is more responsive on resize.

Added the parameter version of the route to App.tsx.

Fixed an explicit any type.

MilesDev91 committed:
Revert variant changes

Changes to variant theme were breaking other pages in the app. This is something that should be done in its own ticket in the future.

MilesDev91 commented:
The requested variant changes ended up breaking other pages. If we have time before the presentation, perhaps we could address this in its own ticket?

rajivtitus commented:
> The requested variant changes ended up breaking other pages. If we have time before the presentation, perhaps we could address this in its own ticket?

That sounds good to me Miles!

rajivtitus reviewed (approved):
Good work with the changes! 💯 
For subsequent PRs please update the PR template with the updated UI images :)

MilesDev91 committed:
Merge branch 'main' into profile-details-page-fe
